server/src/main/java/org/opensearch/index/store/RemoteStoreFileDownloader.java

```diff
@@ -0,0 +1,146 @@
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Helper class to downloads files from a {@link RemoteSegmentStoreDirectory}
+ * instance to a local {@link Directory} instance in parallel depending on thread
+ * pool size and recovery settings.
+ */
+@InternalApi
+public final class RemoteStoreFileDownloader {
+    private final Logger logger;
+    private final ThreadPool threadPool;
+    private final RecoverySettings recoverySettings;
+
+    public RemoteStoreFileDownloader(ShardId shardId, ThreadPool threadPool, RecoverySettings recoverySettings) {
+        this.logger = Loggers.getLogger(RemoteStoreFileDownloader.class, shardId);
+        this.threadPool = threadPool;
+        this.recoverySettings = recoverySettings;
+    }
+
+    /**
+     * Copies the given segments from the remote segment store to the given
+     * local directory.
+     * @param source The remote directory to copy segment files from
+     * @param destination The local directory to copy segment files to
+     * @param toDownloadSegments The list of segment files to download
+     */
+    public void download(Directory source, Directory destination, Collection<String> toDownloadSegments) throws IOException {
+        downloadInternal(source, destination, null, toDownloadSegments, () -> {});
+    }
+
+    /**
+     * Copies the given segments from the remote segment store to the given
+     * local directory, while also copying the segments _to_ another remote directory.
+     * @param source The remote directory to copy segment files from
+     * @param destination The local directory to copy segment files to
+     * @param secondDestination The second remote directory that segment files are
+     *                          copied to after being copied to the local directory
+     * @param toDownloadSegments The list of segment files to download
+     * @param onFileCompletion A generic runnable that is invoked after each file download.
+     *                         Must be thread safe as this may be invoked concurrently from
+     *                         different threads.
+     */
+    public void download(
+        Directory source,
+        Directory destination,
+        Directory secondDestination,
+        Collection<String> toDownloadSegments,
+        Runnable onFileCompletion
+    ) throws IOException {
+        downloadInternal(source, destination, secondDestination, toDownloadSegments, onFileCompletion);
+    }
+
+    private void downloadInternal(
+        Directory source,
+        Directory destination,
+        @Nullable Directory secondDestination,
+        Collection<String> toDownloadSegments,
+        Runnable onFileCompletion
+    ) throws IOException {
+        final Queue<String> queue = new ConcurrentLinkedQueue<>(toDownloadSegments);
```

Comment:
Do you think we could share a common queue for multiple shards that can be recovered concurrently?

Reply:
I think that would be possible if we inject the parallel downloader instance at a high level to be shared across shards. What specifically would you do with that approach? I'm thinking from a monitoring perspective it would be good to know the size of a single queue.